aiida/orm/autogroup.py

```diff
@@ -32,148 +34,240 @@ class Autogroup:
+        )
+
+        if (
+            'all' in include_exact or the_class in include_exact_classes or
+            issubclass(the_class, include_with_subclasses_classes)
         ):
-            exclude = self.get_exclude()
-            exclude_ws = self.get_exclude_with_subclasses()
-            if ((not 'all.' in exclude) or
-                    (the_class._plugin_type_string in exclude) or
-                    any([the_class._plugin_type_string.startswith(i) for i in exclude_ws])
-            ):
+            # According to the include, this class should be included
+            # strings, including possibly 'all'
+            exclude_exact = self.get_exclude()
+            exclude_with_subclasses = self.get_exclude_with_subclasses()
+
+            # actual classes, with 'all' stripped out
+            exclude_exact_classes = tuple(
+                load_entry_point_from_string(ep_string) for ep_string in exclude_exact if ep_string != 'all'
+            )
+            exclude_with_subclasses_classes = tuple(
+                load_entry_point_from_string(ep_string) for ep_string in exclude_with_subclasses if ep_string != 'all'
+            )
+
+            if (the_class not in exclude_exact_classes and not issubclass(the_class, exclude_with_subclasses_classes)):
+                # If we are here, it's not excluded
                 return True
+            # If we're here, it's both in the include and in the exclude - exclude it
+            return False
+        # If we're here, the class is not in the include - return False
+        return False
+
+    def clear_group_cache(self):
+        """Clear the cache of the group name.
+
+        This is mostly used by tests when they reset the database.
+        """
+        self._group_label = None
+
+    def get_or_create_group(self):
+        """Return the current Autogroup, or create one if None has been set yet."""
+        from aiida.orm import QueryBuilder
+        if self._group_label is not None:
+            results = [
+                res[0] for res in QueryBuilder().
+                append(Group, filters={
+                    'label': self._group_label,
+                    'type_string': VERDIAUTOGROUP_TYPE
+                }, project='*').iterall()
+            ]
+            if results:
+                # If it is not empty, it should have only one result due to the
+                # uniqueness constraints
+                return results[0]
+            # There are no results: probably the group has been deleted.
+            # I continue as if it was not cached
+            self._group_label = None
+
+        label_prefix = self.get_group_label_prefix()
+        # Try to do a preliminary QB query to avoid to do too many try/except
+        # if many of the prefix_NUMBER groups already exist
+        queryb = QueryBuilder().append(
+            Group,
+            filters={
+                'or': [{
```

Comment:
The logic is a bit more complex. Say the prefix is `a`. We anyway need to treat specially the case in which no group exists (neither `a` nor `a_%`), the case in which *only* `a` exists, and all other cases. The more complex query is to avoid that we also match labels like `abc`.